File: App/Resources/AlgorithmDetails/redblacktreesort/redblacktreesort.cs
using System;
using System.Collections.Generic;

public class RedBlackTreeSort
{
  private class Node
  {
    public int Value;
    public Node? Left;
    public Node? Right;
    public bool IsRed = true;

    public Node(int value)
    {
      Value = value;
    }
  }

  private class AddResult
  {
    public Node Node;
    public bool NeedsFix;

    public AddResult(Node node, bool needsFix)
    {
      Node = node;
      NeedsFix = needsFix;
    }
  }

  private static bool IsRed(Node? node) => node != null && node.IsRed;

  private static Node SingleRotateRight(Node node)
  {
    Node b = node.Left!;
    node.Left = b.Right;
    b.Right = node;
    b.IsRed = false;
    node.IsRed = true;
    return b;
  }

  private static Node SingleRotateLeft(Node node)
  {
    Node b = node.Right!;
    node.Right = b.Left;
    b.Left = node;
    b.IsRed = false;
    node.IsRed = true;
    return b;
  }

  private static Node DoubleRotateRight(Node node)
  {
    node.Left = SingleRotateLeft(node.Left!);
    return SingleRotateRight(node);
  }

  private static Node DoubleRotateLeft(Node node)
  {
    node.Right = SingleRotateRight(node.Right!);
    return SingleRotateLeft(node);
  }

  private static AddResult Add(Node? node, int value)
  {
    if (node == null)
    {
      return new AddResult(new Node(value), false);
    }

    if (!node.IsRed && IsRed(node.Left) && IsRed(node.Right))
    {
      node.IsRed = true;
      node.Left!.IsRed = false;
      node.Right!.IsRed = false;
    }

    if (value < node.Value)
    {
      AddResult child = Add(node.Left, value);
      node.Left = child.Node;
      if (child.NeedsFix)
      {
        if (IsRed(node.Left!.Left))
        {
          return new AddResult(SingleRotateRight(node), false);
        }
        return new AddResult(DoubleRotateRight(node), false);
      }
      return new AddResult(node, node.IsRed && IsRed(node.Left));
    }
    else
    {
      AddResult child = Add(node.Right, value);
      node.Right = child.Node;
      if (child.NeedsFix)
      {
        if (IsRed(node.Right!.Right))
        {
          return new AddResult(SingleRotateLeft(node), false);
        }
        return new AddResult(DoubleRotateLeft(node), false);
      }
      return new AddResult(node, node.IsRed && IsRed(node.Right));
    }
  }

  private static void Traverse(Node? node, List<int> result)
  {
    if (node == null)
      return;
    Traverse(node.Left, result);
    result.Add(node.Value);
    Traverse(node.Right, result);
  }

  public static void Sort(int[] arr)
  {
    int n = arr.Length;
    Node? root = null;
    for (int i = 0; i < n; i++)
    {
      AddResult inserted = Add(root, arr[i]);
      root = inserted.Node;
      root.IsRed = false;
    }

    List<int> result = new List<int>();
    Traverse(root, result);

    for (int i = 0; i < n; i++)
    {
      arr[i] = result[i];
    }
  }

  public static void Main(String[] args)
  {
    int[] array = { 0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56 };
    Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}

File: App/Resources/AlgorithmDetails/redblacktreesort/redblacktreesort.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
    this.isRed = true;
  }
}

function isRed(node) {
  return node !== null && node.isRed;
}

function singleRotateRight(node) {
  const b = node.left;
  node.left = b.right;
  b.right = node;
  b.isRed = false;
  node.isRed = true;
  return b;
}

function singleRotateLeft(node) {
  const b = node.right;
  node.right = b.left;
  b.left = node;
  b.isRed = false;
  node.isRed = true;
  return b;
}

function doubleRotateRight(node) {
  node.left = singleRotateLeft(node.left);
  return singleRotateRight(node);
}

function doubleRotateLeft(node) {
  node.right = singleRotateRight(node.right);
  return singleRotateLeft(node);
}

function add(node, value) {
  if (node === null) {
    return { node: new Node(value), needsFix: false };
  }

  if (!node.isRed && isRed(node.left) && isRed(node.right)) {
    node.isRed = true;
    node.left.isRed = false;
    node.right.isRed = false;
  }

  if (value < node.value) {
    const child = add(node.left, value);
    node.left = child.node;
    if (child.needsFix) {
      if (isRed(node.left.left)) {
        return { node: singleRotateRight(node), needsFix: false };
      }
      return { node: doubleRotateRight(node), needsFix: false };
    }
    return { node: node, needsFix: node.isRed && isRed(node.left) };
  }

  const child = add(node.right, value);
  node.right = child.node;
  if (child.needsFix) {
    if (isRed(node.right.right)) {
      return { node: singleRotateLeft(node), needsFix: false };
    }
    return { node: doubleRotateLeft(node), needsFix: false };
  }
  return { node: node, needsFix: node.isRed && isRed(node.right) };
}

function traverse(node, result) {
  if (node === null) {
    return;
  }
  traverse(node.left, result);
  result.push(node.value);
  traverse(node.right, result);
}

function sort(arr) {
  const n = arr.length;
  let root = null;
  for (let i = 0; i < n; i++) {
    const inserted = add(root, arr[i]);
    root = inserted.node;
    root.isRed = false;
  }

  const result = [];
  traverse(root, result);

  for (let i = 0; i < n; i++) {
    arr[i] = result[i];
  }
  return arr;
}

var array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56];
sort(array);
console.log("[" + array.join(", ") + "]");
